### PR TITLE
Migrate/user settings modal

### PR DESCRIPTION
**Pull request by oscarcow1**
_Any italic text should be deleted from the final Pull Request text, including this line_

# Description

Migrate user-settings-modal to typescript, also make the page can slip.

## Type of change

_Please delete options that are not relevant._

- [x] Component Migration

# How Has This Been Tested?

Can successfully run in Chrome, but still need npm test to test in the other browsers.

## Testing Checklist:

- [x] Tested in latest Chrome
- [ ] Tested in latest Safari
- [ ] Tested in latest Firefox

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have requested a review from  @RayGuo-ergou, @PerryRose and @tancnle on the Pull Request

### CODE REVIEW

**Comment by oscarcow1** [comment]
[User-settings-modal migration](https://trello.com/c/IcJ5SMSp/45-user-settings-modal-migration)

File: src/app/admin/modals/user-settings-modal/user-settings-modal.component.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
+import { Component, OnInit, Inject } from '@angular/core';
+import { MatDialogRef, MAT_DIALOG_DATA } from '@angular/material/dialog';
+import { currentUser } from 'src/app/ajs-upgraded-providers';
+import { User, UserService } from 'src/app/api/models/doubtfire-model';
+import { DoubtfireConstants } from 'src/app/config/constants/doubtfire-constants';
+
+@Component({
+  selector: 'user-settings-modal',
```
Comment (RayGuo-ergou) [inline src/app/admin/modals/user-settings-modal/user-settings-modal.component.ts:8]: as the [CI](https://github.com/thoth-tech/doubtfire-web/runs/6115326969?check_suite_focus=true) mentioned : The selector should start with one of these prefixes: "df". Can we change that?

File: src/app/admin/modals/user-settings-modal/user-settings-modal.component.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    // edit detail
+    if (this.data.id) {
+      this.isNew = false;
+      // send user data
+      this.user = this.data;
+
+      // or search basic on ID
+      // this.userService.query({ id: this.data.id }).subscribe((data) => {
+      //   this.user = data[0]
+      // })
```
Comment (RayGuo-ergou) [inline src/app/admin/modals/user-settings-modal/user-settings-modal.component.ts:42]: **Suggestion**: Can we remove the unrelated code?

File: src/app/admin/modals/user-settings-modal/user-settings-modal.component.html [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    <div class="form-group" *ngIf="isNew">
+      <label class="col-sm-3 control-label" ng-required="isNew">Username</label>
+      <div class="col-sm-7">
+        <input [(ngModel)]="user.username" type="input" class="form-control" placeholder="1744070" />
+      </div>
+    </div>
+    <div class="form-group" ng-required="isNew">
+      <label class="col-sm-3 control-label">First Name</label>
+      <div class="col-sm-7">
+        <input [(ngModel)]="user.first_name" type="input" class="form-control"
```
Comment (RayGuo-ergou) [inline src/app/admin/modals/user-settings-modal/user-settings-modal.component.html:21]: There is an error when loading this modal:

<details>
  <summary>ERROR TypeError: Cannot read properties of undefined (reading 'first_name')</summary>

at UserSettingsModalComponent_Template (user-settings-modal.component.html:21:16)

at executeTemplate (core.mjs:9603:1)

at refreshView (core.mjs:9469:1)

at refreshComponent (core.mjs:10640:1)

at refreshChildComponents (core.mjs:9265:1)

at refreshView (core.mjs:9519:1)

at refreshEmbeddedViews (core.mjs:10594:1)

at refreshView (core.mjs:9493:1)

at refreshComponent (core.mjs:10640:1)

at refreshChildComponents (core.mjs:9265:1)
</details>

File: src/app/admin/modals/user-settings-modal/user-settings-modal.component.html [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        <label class="col-sm-3 control-label">Allow Analytics</label>
+        <div class="col-sm-7">
+          <!-- <div class="btn-group">
+            <label class="btn btn-default" [(ngModel)]="user.opt_in_to_research" btn-radio="true">
+              <i class="fa fa-check"></i>
+            </label>
+            <label class="btn btn-default" [(ngModel)]="user.opt_in_to_research" btn-radio="false">
+              <i class="fa fa-close"></i>
+            </label>
+          </div> -->
```
Comment (RayGuo-ergou) [inline src/app/admin/modals/user-settings-modal/user-settings-modal.component.html:102]: **Suggestion**: Can we remove the unrelated code?

File: src/app/admin/modals/user-settings-modal/user-settings-modal.component.html [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+          <mat-checkbox class="example-margin" [(ngModel)]="user.opt_in_to_research">Checked</mat-checkbox>
+          <p class="help-block">Allow us to anonymously gather your usage to improve {{externalName }}</p>
+        </div>
+      </div>
+    </div>
+  </div>
+  <div class="modal-footer text-right">
+    <input (click)="saveUser()" class="btn btn-primary" value="{{isNew ? 'Add' : 'Save'}}" />
+  </div>
```
Comment (RayGuo-ergou) [inline src/app/admin/modals/user-settings-modal/user-settings-modal.component.html:111]: The button has no text and does not work due to ` ERROR TypeError: Cannot read properties of undefined (reading 'first_name')`
image

File: src/app/admin/modals/user-settings-modal/user-settings-modal.component.html [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+          </mat-radio-button>
+          <mat-radio-button value="Student">
+            Student
+          </mat-radio-button>
+        </mat-radio-group>
+        <p *ngIf="user.id === CurrentUser.id" class="help-block text-center">
+          You cannot modify your own permissions
+        </p>
+      </div>
+    </div>
```
Comment (RayGuo-ergou) [inline src/app/admin/modals/user-settings-modal/user-settings-modal.component.html:89]: This section does not work due to `ERROR TypeError: Cannot read properties of undefined (reading 'first_name')`

File: src/app/admin/modals/user-settings-modal/user-settings-modal.component.ts [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+  public updateExistingUser() {
+    const { id, first_name, last_name } = this.user;
+    this.userService.update(this.currentWebcalWith(this.user)).subscribe(response => {
+      this.userService.save(response)
+      this.dialogRef.close();
+    })
+  }
+
+  // show example on table
+  private currentWebcalWith(o: Partial<User>): User {
```
Comment (RayGuo-ergou) [inline src/app/admin/modals/user-settings-modal/user-settings-modal.component.ts:65]: Can we have a resonable name instead of `o` ?

**Review by RayGuo-ergou — CHANGES_REQUESTED** [review]
Hi @oscarcow1 Thanks for your hard work. I have quickly reviewed and runned your code. However, there is an error when open the user setting modal, could you fix it? And also please create a unit test for you component and also running the code formatter by ` npx prettier --write <filename>` or useing the intergrated VSCode extension.

File: src/app/admin/modals/user-settings-modal/user-settings-modal.component.ts [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+  }
+
+  // create new user
+  public createNewUser() {
+    this.userService.create({
+      user: this.user
+    }).subscribe(response => {
+      this.dialogRef.close();
+    })
+  }
```
Comment (RayGuo-ergou) [inline src/app/admin/modals/user-settings-modal/user-settings-modal.component.ts:53]: **Suggestion**: can we have error control for this call? see https://rxjs.dev/deprecations/subscribe-arguments

File: src/app/admin/modals/user-settings-modal/user-settings-modal.component.ts [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+  // user edit
+  public updateExistingUser() {
+    const { id, first_name, last_name } = this.user;
+    this.userService.update(this.currentWebcalWith(this.user)).subscribe(response => {
+      this.userService.save(response)
+      this.dialogRef.close();
+    })
+  }
```
Comment (RayGuo-ergou) [inline src/app/admin/modals/user-settings-modal/user-settings-modal.component.ts:62]: **Suggestion**: can we have error control for this call? see https://rxjs.dev/deprecations/subscribe-arguments

File: src/app/admin/modals/user-settings-modal/user-settings-modal.component.ts (hunk 7 above)
Comment (RayGuo-ergou) [inline src/app/admin/modals/user-settings-modal/user-settings-modal.component.ts:65]: **Question**: Do we need `Partial` for user? I think it's used for updating the object beceuse it returns a type that represents all subsets of a given type. However, I did not see any modifacation after you define Partial object.

File: src/app/admin/modals/user-settings-modal/user-settings-modal.component.ts [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+    @Inject(currentUser) public CurrentUser: any,
+    public dialogRef: MatDialogRef<UserSettingsModalComponent>,
+    private constants: DoubtfireConstants,
+    private userService: UserService
+  ) {}
+
+  ngOnInit(): void {
+    this.constants.ExternalName.subscribe((result) => {
+      this.externalName = result;
+    })
```
Comment (RayGuo-ergou) [inline src/app/admin/modals/user-settings-modal/user-settings-modal.component.ts:31]: **Question**: Do we need `subscribe`? Isn't the ` ExternalName ` directly from ` DoubtfireConstants.ExternalName `?

**Comment by RayGuo-ergou** [comment]
Update: did more test, seems the edit user is fine the issue is create new user.

**Comment by oscarcow1** [comment]
Thanks for your time, will fix this in this weekend

**Comment by oscarcow1** [comment]
The add button can be use now, but it need to refresh the page to update the added user, looks like need to change user.coffee file, but I am afraid I can not fix this in this moment.

**Comment by RayGuo-ergou** [comment]
Can we discuss about my first review suggestions before I review the newest changes?

**Comment by oscarcow1** [comment]
> Can we discuss about my first review suggestions before I review the newest changes?

Hi Ray, what things do you want to discuss further?  I am sorry to say that I didn't follow your suggestions like format style, cause I want to make sure the function is work, but I will look about this later

File: package.json [hunk 11]
```diff
@@ -12,7 +12,7 @@
     "build": "run-s -l build:angular1 build:angular13",
     "build:angular1": "grunt build",
     "build:angular13": "ng build",
-    "serve:angular13": "export NODE_OPTIONS=--max_old_space_size=4096 && ng serve --host 0.0.0.0",
+    "serve:angular13": "set NODE_OPTIONS=--max_old_space_size=8000 && ng serve --host 0.0.0.0",
```
Comment (tancnle) [inline package.json:15]: **question:** I am curious of why this command is changed as part of this PR? 🤔

**Comment by RayGuo-ergou** [comment]
> cause I want to make sure the function is work, but I will look about this later

I think we can do the debug and other things together. and had a quick look at your commit. I think you might over thinking this bug, should be a esay fix by adding a if statement in the HTML like ` {{ user? user: '' }}`. So my recommandation would be revert this commit, you can find how to do it in this [post](https://stackoverflow.com/questions/4114095/how-do-i-revert-a-git-repository-to-a-previous-commit). Then we can resolve things one by one.

File: package.json (hunk 11 above)
Comment (RayGuo-ergou) [reply package.json:15]: > **question:** I am curious of why this command is changed as part of this PR? 🤔

I think it because he's using windows, and commit this file by accident.

Comment (oscarcow1) [reply package.json:15]: Cause I am using Windows, If I use "export", I cannot run doubt-fire locally, means I cannot hot update by any changes, so I need this command, this method is helped by Ray.

Comment (RayGuo-ergou) [reply package.json:15]: > Cause I am using Windows, If I use "export", I cannot run doubt-fire locally, means I cannot hot update by any changes, so I need this command, this method is helped by Ray.

This kind of changes should not push to github, so it would be great that you don't use ` git add .` to add all files. You can add the files need to be push to github one by one using command `git add <filename>` or use VScode git control panel (stage change means git add).

Comment (tancnle) [reply package.json:15]: Would this be problematic for non-windows users? How about an alternative that could work in both systems? 🤔 

```js
"serve:angular13": "NODE_OPTIONS=--max_old_space_size=8000 ng serve --host 0.0.0.0",
```

Comment (RayGuo-ergou) [reply package.json:15]: > Would this be problematic for non-windows users? How about an alternative that could work in both systems? 🤔
> 
> ```js
> "serve:angular13": "NODE_OPTIONS=--max_old_space_size=8000 ng serve --host 0.0.0.0",
> ```

Yes, this would be an error like:
image

Comment (tancnle) [reply package.json:15]: It would also be nice to address this issue in a separate PR.

**Comment by oscarcow1** [comment]
> > cause I want to make sure the function is work, but I will look about this later
> 
> I think we can do the debug and other things together. and had a quick look at your commit. I think you might over thinking this bug, should be a esay fix by adding a if statement in the HTML like ` {{ user? user: '' }}`. So my recommandation would be revert this commit, you can find how to do it in this [post](https://stackoverflow.com/questions/4114095/how-do-i-revert-a-git-repository-to-a-previous-commit). Then we can resolve things one by one.

> Thank you Ray, sorry for my poor skills, I might not over thinking, just lack of the knowledge to fix this problem, I need to research so many sources to solve this problem.. Its hard for me to understand your thinking on this component, can you please give me more suggestions on it and I will take my time on this project further. Thank you

**Comment by RayGuo-ergou** [comment]
>The add button can be use now, but it need to refresh the page to update the added user,

It okay, you are doing the migration so just stay its functionality. But the update user feature should have the live update, just make sure about that.

**Comment by RayGuo-ergou** [comment]
>can you please give me more suggestions on it and I will take my time on this project further.

Maybe we could have a video call to discuess? What do you think?

**Comment by oscarcow1** [comment]
will DM you before the video call when you'refree, thank you

获取Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by RayGuo-ergou** [comment]
@oscarcow1 as discuessed I pushed a commit to refactor the code and if you have any question just leave a commnet here or message me on Teams.

I haven't done the migration from boostrap to Material, so you still need to do it. You can follow @PerryRose 's suggestion like making a  `<mat-card>` or just change the class from boostrap to Material. I think both is fine. Before you do that dot forget to get my update by using command 
```
git checkout migrate/user-settings-modal

git pull
```

File: src/app/admin/modals/user-settings-modal/user-settings-modal.component.ts (hunk 8 above)
Comment (zhaoyiq0719) [reply src/app/admin/modals/user-settings-modal/user-settings-modal.component.ts]: Response has been defined, why is it not used?

File: src/app/admin/modals/user-settings-modal/user-settings-modal.component.ts [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+  // create or not
+  isNew: boolean = true;
+  // user information list
+  user: User | null;
+
+  externalName: string;
+
+  constructor(
+    @Inject(MAT_DIALOG_DATA) public data: any,
+    @Inject(currentUser) public CurrentUser: any,
```
Comment (zhaoyiq0719) [inline src/app/admin/modals/user-settings-modal/user-settings-modal.component.ts:20]: Can we use user objects instead of any?

File: src/app/admin/modals/user-settings-modal/user-settings-modal.component.ts [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+      user: this.user
+    }).subscribe(response => {
+      this.dialogRef.close();
+    })
+  }
+
+  // user edit
+  public updateExistingUser() {
+    const { id, first_name, last_name } = this.user;
+    this.userService.update(this.currentWebcalWith(this.user)).subscribe(response => {
```
Comment (zhaoyiq0719) [inline src/app/admin/modals/user-settings-modal/user-settings-modal.component.ts]: ID, first name, last name not used. Can we cancalled them?

**Comment by oscarcow1** [comment]
> @oscarcow1 as discuessed I pushed a commit to refactor the code and if you have any question just leave a commnet here or message me on Teams.
> 
> I haven't done the migration from boostrap to Material, so you still need to do it. You can follow @PerryRose 's suggestion like making a `<mat-card>` or just change the class from boostrap to Material. I think both is fine. Before you do that dot forget to get my update by using command
> 
> ```
> git checkout migrate/user-settings-modal
> 
> git pull
> ```
> 
> Also for the **src/app/admin/modals/user-settings-modal/user-settings-modal.services.ts**, i perfer `Observables ` over ` promise`. Since you made it `promise` so I did not change it. But you can change it to `observables` when you do the UI migration. But I think its also fine to leave it as `promise` so its your decision.

Thanks for your time Ray, will follow your suggestion and make changes, appreciate it.

**Comment by oscarcow1** [comment]
Boostrap to Material migration is finished, but the set type might not the best, please have a check and give me feedback, and I will make changes. Also use Observable instead of Promise, and use objects instead of any. Happy to having any comments, will reply ASAP, thanks a lot:)

**Comment by RayGuo-ergou** [comment]
Thanks @oscarcow1 for your update, however, I still can see some Bootstrap class for example ` class="col-sm-9"`,  ` class="help-block"` and ` class="btn btn-sm btn-default"`. Be careful of every HTML class attribute. I think after this minor change, we are ready to go.👍

File: src/app/admin/states/users/users.coffee [hunk 14]
```diff
@@ -44,7 +44,11 @@ angular.module('doubtfire.admin.states.users', [])
   $scope.showUserModal = (user) ->
     # If we're given a user, show that user, else create a new one
     userToShow = if user? then user else new User { }
-    UserSettingsModal.show userToShow
+    UserSettingsModal.show(userToShow).subscribe (result) ->
+      if result.action is "create"
+        $scope.users = $scope.users.concat(result.user)
+      else if !result.action
```
Comment (RayGuo-ergou) [inline src/app/admin/states/users/users.coffee:50]:
```suggestion
      else
```

File: src/app/admin/modals/user-settings-modal/user-settings-modal.services.ts [hunk 15]
```diff
@@ -8,21 +10,20 @@ import { UserSettingsModalComponent } from './user-settings-modal.component';
...
-        // create/update successful resolve
-        // else reject with info message
         if (result) {
-          resolve(result);
+          subscriber.next(result);
         } else {
-          reject('Cancelled Update/Create');
+          subscriber.next('Cancelled Update/Create');
         }
+        subscriber.complete();
```
Comment (RayGuo-ergou) [inline src/app/admin/modals/user-settings-modal/user-settings-modal.services.ts:26]: Logic not correct. For promise used `resolve` and ` reject` notice they are differnt keywords.

File: src/app/admin/modals/user-settings-modal/user-settings-modal.services.ts [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+      let dialogRef: MatDialogRef<UserSettingsModalComponent>;
+      dialogRef = this.dialog.open(UserSettingsModalComponent, {
+        data: user,
+      });
+      dialogRef.afterClosed().subscribe((result) => {
+        if (result) {
+          subscriber.next(result);
+        } else {
+          subscriber.next('Cancelled Update/Create');
+        }
```
Comment (RayGuo-ergou) [inline src/app/admin/modals/user-settings-modal/user-settings-modal.services.ts:25]:
```suggestion
      subscriber.next(result);
```

File: src/app/admin/modals/user-settings-modal/user-settings-modal.services.ts [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+        subscriber.complete();
```
Comment (RayGuo-ergou) [inline src/app/admin/modals/user-settings-modal/user-settings-modal.services.ts:26]:
```suggestion

```
Should keep listen

File: src/app/admin/states/users/users.coffee [hunk 18]
```diff
@@ -44,7 +44,11 @@ angular.module('doubtfire.admin.states.users', [])
...
+        alertService.add "info", result, 2000
```
Comment (RayGuo-ergou) [inline src/app/admin/states/users/users.coffee:51]:
```suggestion
      if result.action is "create"
        $scope.users = $scope.users.concat(result.user)
      else 
        alertService.add "info", "Cancelled Update/Create", 2000
```

**Review by RayGuo-ergou — COMMENTED** [review]
@oscarcow1 Leaved some suggestion on Observable.

File: src/app/admin/states/users/users.coffee (hunk 18 above)
Comment (RayGuo-ergou) [reply src/app/admin/states/users/users.coffee:51]: Also an error handler would be great

Comment (RayGuo-ergou) [inline src/app/admin/states/users/users.coffee:51]:
```suggestion
      alertService.add "info", "'Cancelled Update/Create'", 2000
```

Comment (RayGuo-ergou) [inline src/app/admin/states/users/users.coffee:51]:
```suggestion
      if result is undefined
        alertService.add "info", "Cancelled Update/Create", 2000
      else if result.action is "create"
        $scope.users = $scope.users.concat(result.user)
```
